pkg/syncer/k8scloudoperator/placement.go

```diff
@@ -410,6 +419,204 @@ func stampPVCWithError(ctx context.Context, client kubernetes.Interface,
+func eliminateNodesWithPvcOfSiblingReplica(ctx context.Context, client kubernetes.Interface, currPVC *v1.PersistentVolumeClaim, candidateHosts []string) ([]string, error) {
+	log := logger.GetLogger(ctx)
+
+	containerOrchestratorUtility := commonco.ContainerOrchestratorUtility
+	if containerOrchestratorUtility == nil {
+		clusterFlavor, err := cnsconfig.GetClusterFlavor(ctx)
+		if err != nil {
+			log.Debugf("Failed retrieving cluster flavor. Error: %v", err)
+			return candidateHosts, nil
+		}
+		containerOrchestratorUtility, err = commonco.GetContainerOrchestratorInterface(ctx, common.Kubernetes, clusterFlavor, *admissionhandler.COInitParams)
+		if err != nil {
+			log.Debugf("failed to get k8s interface. err: %v", err)
+			return candidateHosts, nil
+		}
+	}
+
+	if !containerOrchestratorUtility.IsFSSEnabled(ctx, common.SiblingReplicaBoundPvcCheck) {
+		log.Infof("FSS to check for sibling replica's PVCs is not enabled.")
+		return candidateHosts, nil
+	}
+
+	// Proceed only if it is a vDPP PVC
+	pvcLabels := currPVC.GetLabels()
+	appplatformLabelVal, labelOk := pvcLabels[appplatformLabel]
+	if !labelOk {
+		log.Infof("Not a vDPP PVC %+v, skip this step.", currPVC.Name)
+		return candidateHosts, nil
+	}
+
+	// Do not proceed if it is a standalone pod
+	pvcAnnotation := currPVC.ObjectMeta.Annotations
+	if _, ok := pvcAnnotation[nodeAffinityAnnotationKey]; ok {
+		log.Infof("Node affinity key already specified for PVC %+v, skip this step.", currPVC.Name)
+		return candidateHosts, nil
+	}
+
+	cousinPVCs, err := getCousinPVCs(ctx, client, currPVC)
+	if err != nil {
+		return candidateHosts, err
+	}
+
+	selectLabel := fmt.Sprintf("%s=%s", appplatformLabel, appplatformLabelVal)
+	allPVCs, err := client.CoreV1().PersistentVolumeClaims(currPVC.Namespace).List(ctx, metav1.ListOptions{
+		LabelSelector: selectLabel,
+	})
+	if err != nil {
+		return candidateHosts, err
+	}
+
+	for _, cousinPVC := range cousinPVCs {
+		pvc := getPVCObject(cousinPVC, allPVCs.Items)
+		if pvc == nil {
+			// This scenario can happen if psp-operator PVC watcher has already deleted the PVC
+			log.Errorf("Failed to find PVC object for PVC %s", cousinPVC)
+			continue
+		}
+
+		pvName := pvc.Spec.VolumeName
+		pv, err := client.CoreV1().PersistentVolumes().Get(ctx, pvName, metav1.GetOptions{})
+		if err != nil || pv == nil {
+			if pvName == "" {
+				continue
+			}
+			return candidateHosts, err
+		}
+
+		// Volume Binding has to be WaitForFirstConsumer because it can only be Immediate if it has the annotation failure-domain.beta.vmware.com/node on the PVC.
+		// We have already verified that this annotation is not present on PVC as part of the validatin above.
+		// Hence, nodeAffinity value will always be defined on the PV.
+		hostName := pv.Spec.NodeAffinity.Required.NodeSelectorTerms[0].MatchExpressions[0].Values[0]
+
+		for i, host := range candidateHosts {
+			if host == hostName {
+				candidateHosts = filterHost(candidateHosts, i)
```

Comment:
Can you verify this for loop block by either writing unit test or a sample code? I am asking because we are changing the array `candidateHosts` while looping over it.

Reply:
As the loop is breaking as soon as I am finding a match, it is not a problem.
I have written a test code here: https://play.golang.org/p/daAchiOnf1J